### PR TITLE
Minor update to make llv2_encryption_utility_test less flaky.

### PR DESCRIPTION
**Pull request by wangyaopw**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108)

### CODE REVIEW

**Comment by wangyaopw** [comment]
So what these two tests are doing is that we add some noises to the protocol and we make sure the randomness of the noise are expected.

The distribution of the random number is tested somewhere else in the random number generator test.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108)* status: all files reviewed, 1 unresolved discussion (waiting on @wangyaopw and @zachcwc)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108#-McGxx002GVw5hdBVi-q:-McGxx002GVw5hdBVi-r:bcx4nr3) (no related file):*
Approving this because it's still an improvement, but ideally we should have a way to make this not flaky at all by being able to specify the seed for the random number generator.

---

**Comment by zachcwc** [comment]
___
*[src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc, line 1048 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108#-McLC3eG4ukDXaBABp7b:-McLC3eH7D9XNprKFH1P:b-d88v7g) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7bf3c83189725f440c9a0ff7b1075a86c43e9273/src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc#L1048)):*
> ```cpp
>   frequency_noise_params.set_maximum_frequency(30);
>   frequency_noise_params.set_contributors_count(kWorkerCount);
>   // resulted p = 0.606531, offset = 9
> ```

How is the offset determined again? I can't find it in the liquid legion sketch generation docs

**Review by wangyaopw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108)* status: all files reviewed, 2 unresolved discussions (waiting on @wangyaopw and @zachcwc)

___
*[src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc, line 1048 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108#-McLC3eG4ukDXaBABp7b:-McLKfMGB3cxp1Z4Ae1r:b76yl56) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7bf3c83189725f440c9a0ff7b1075a86c43e9273/src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc#L1048)):*
<details><summary><i>Previously, zachcwc wrote…</i></summary><blockquote>

How is the offset determined again? I can't find it in the liquid legion sketch generation docs
</blockquote></details>

The noise algorithm is roughly described in the doc "[Draft for WFA] An Improved MPC Protocol for Reach and Frequency Computation". There is a more detail academic paper for this part (not yet published.

tldr is that each worker adds a certain amount noise, and the total number of noises added by all workers is bidirectional geometric distributed.

the values used in the test are simply obtained from running the code in https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/main/src/main/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/noise_parameters_computation.cc#L80

**Comment by zachcwc** [comment]
___
*[src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc, line 1048 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108#-McLC3eG4ukDXaBABp7b:-McLN4cI8kCcCmtEnVLC:bslgevp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7bf3c83189725f440c9a0ff7b1075a86c43e9273/src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc#L1048)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

The noise algorithm is roughly described in the doc "[Draft for WFA] An Improved MPC Protocol for Reach and Frequency Computation". There is a more detail academic paper for this part (not yet published.

tldr is that each worker adds a certain amount noise, and the total number of noises added by all workers is bidirectional geometric distributed.

the values used in the test are simply obtained from running the code in https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/main/src/main/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/noise_parameters_computation.cc#L80

</blockquote></details>

Yeah I have that doc, but I don't see "offset" in it. From the code you sent I gather that it is the computed polya variable?

**Review by wangyaopw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108)* status: all files reviewed, 2 unresolved discussions (waiting on @wangyaopw and @zachcwc)

___
*[src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc, line 1048 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108#-McLC3eG4ukDXaBABp7b:-McLNn7M70oRk0uJcA1M:bfus6gy) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7bf3c83189725f440c9a0ff7b1075a86c43e9273/src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc#L1048)):*
<details><summary><i>Previously, zachcwc wrote…</i></summary><blockquote>

Yeah I have that doc, but I don't see "offset" in it. From the code you sent I gather that it is the computed polya variable?
</blockquote></details>

Yep. The polya variable is truncated and shifted.

The original polya distribution is centered at 0. We first truncate it to the range of [-offset, offset], and then shift it to [0, 2*offset], such that the value is always non-negative, since the noise register count should be non-negative.

When we do the estimation, we subtract "offset" from the result.

**Comment by zachcwc** [comment]
___
*[src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc, line 1048 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108#-McLC3eG4ukDXaBABp7b:-McLPWdg-hD3jEQxzGd9:b-owjxeo) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/7bf3c83189725f440c9a0ff7b1075a86c43e9273/src/test/cc/wfa/measurement/internal/duchy/protocol/liquid_legions_v2/liquid_legions_v2_encryption_utility_test.cc#L1048)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

Yep. The polya variable is truncated and shifted.

The original polya distribution is centered at 0. We first truncate it to the range of [-offset, offset], and then shift it to [0, 2*offset], such that the value is always non-negative, since the noise register count should be non-negative.

When we do the estimation, we subtract "offset" from the result.
</blockquote></details>

Ah gotcha ok thank you!

**Review by zachcwc — APPROVED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/108)* status: all files reviewed, 1 unresolved discussion (waiting on @wangyaopw)
